clang-tools-extra/test/clang-tidy/checkers/bsl-template-generic-param.cpp

```diff
@@ -0,0 +1,161 @@
+// RUN: %check_clang_tidy %s bsl-template-generic-param %t
+
+#include <cstdint>
+
+// simplified
+class A {
+public:
+  // template constructor like copy constructor
+  template <typename T> A(T const &rhs) {}
+  // CHECK-MESSAGES: [[@LINE-1]]:2: warning: a copy constructor should be declared when there is a template constructor with a single parameter that is a generic parameter. [bsl-template-generic-param]
+
+  // like copy assignment
+  template <typename T> T &operator=(T const &rhs) { return *this; }
+  // CHECK-MESSAGES: [[@LINE-1]]:6: warning: a copy assignment operator should be declared when there is a template assignment operator with a parameter that is a generic parameter. [bsl-template-generic-param]
+};
+
+class B {
+public:
+  // copy constructor
+  B(const B &old_obj) {}
+  // copy assignment
+  B &operator=(B other) {} // (const B &other)
+
+  template <typename T> B(T const &rhs) {}
+
+  template <typename T> T &operator=(T const &rhs) { return *this; }
+};
+
+// og
+// class A
+// {
+// public:
+// 	A ( );
+// 	// A ( A const & rhs );
+// 	// like copy constructor
+// 	template <typename T>
+// 	A ( T const & rhs ): i ( new int )
+// 	{
+// 		*i = *rhs.i;
+// 	}
+
+// 	// like copy assignment
+// 	template <typename T>
+// 	T & operator= ( T const & rhs ) // Example 2
+// 	{
+// 		if ( this != &rhs ) {
+// 			delete i;
+// 			i = new int32_t;
+// 			*i = *rhs.i;
+// 		}
+// 		return *this;
+// 	}
+// 	private:
+// 		int * i;
+// };
+
+// void f ( A const & a1 )		// remove from test
+// {
+// 	A a2 ( a1 );
+// }
+
+// class B
+// {
+// public:
+// 	// copy constructor
+// 	B (const B &old_obj) {}
+// 	// copy assignment
+// 	B& operator=(B other) {}	// (const B &other)
+
+// 	// A & operator= ( A const & rhs )	// Example 1 - implicitly
+// generated
+// 	// {
+// 		// i = rhs.i;
+// 		// return *this;
+// 	// }
+// 	template <typename T>
+// 	B ( T const & rhs ): i ( new int )
+// 	{
+// 		*i = *rhs.i;
+// 	}
+
+// 	template <typename T>
+// 	T & operator= ( T const & rhs ) // Example 2
+// 	{
+// 		if ( this != &rhs ) {
+// 			delete i;
+// 			i = new int32_t;
+// 			*i = *rhs.i;
+// 		}
+// 		return *this;
+// 	}
+// private:
+// 	int32_t * i; // Member requires deep copy
+// // };
+// void f2 ( B const & b1, B & b2 )		// remove
+// {
+// 	b2 = b1; // Unexpectedly uses Example 1
+// }
+
+// additional
+template <typename T> class C {
+public:
+  // template constructor like copy constructor
+  template <typename Y> C(T const &rhs) {}
+
+  // like copy assignment
+  C &operator=(T const &rhs) { return *this; }
+};
+
+class D {
+public:
+  template <typename T> D(T myvar) {}
+  // CHECK-MESSAGES: [[@LINE-1]]:3: warning: a copy constructor should be declared when there is a template constructor with a single parameter that is a generic parameter. [bsl-template-generic-param]
+};
+
+template <typename T> class E {
+public:
+  template <typename Y> E(T myvar2) {}
```

Comment:
Should this warn?

Reply:
The template type in the argument is not the same as the one defined in the function template (Y vs T), so my understanding is that this shouldn't match?